cdc/redo/writer/writer.go

```diff
@@ -555,19 +521,71 @@ func (l *LogWriter) getMetafileName() string {
 		common.DefaultMetaFileType, common.MetaEXT)
 }
 
-func (l *LogWriter) flushLogMeta(checkPointTs, resolvedTs uint64) error {
+func (l *LogWriter) mergeMeta(checkPointTs uint64, rtsMap map[model.TableID]model.Ts) ([]byte, error) {
 	l.metaLock.Lock()
 	defer l.metaLock.Unlock()
 
-	if checkPointTs != 0 {
+	hasChange := false
+	if checkPointTs > l.meta.CheckPointTs {
 		l.meta.CheckPointTs = checkPointTs
+		hasChange = true
+	} else if checkPointTs > 0 && checkPointTs != l.meta.CheckPointTs {
+		log.Panic("flushLogMeta with a regressed checkpoint ts",
+			zap.Uint64("currCheckPointTs", l.meta.CheckPointTs),
+			zap.Uint64("recvCheckPointTs", checkPointTs))
+	}
+
+	for tID, ts := range rtsMap {
+		if l.meta.ResolvedTsList[tID] == ts {
```

Comment:
Is it assumed that `rtsMap` contains all keys in `l.meta.ResolvedTsList`? Do we need to check the existence of `l.meta.ResolvedTsList[tID]`?

Reply:
nice catch!

Reply:
Oh, if `tID` doesn't exist in `l.meta.ResolvedTsList`, `l.meta.ResolvedTsList[tID]` will be 0. So seems it's still ok.